Init Openshift documentation (#1315)
elastic/cloud-on-k8s: docs/openshift.asciidoc

[id="{p}-openshift"]

> == Openshift

Reviewer: `s|Openshift|OpenShift|g`: I think the correct letter case is Open**S**hift.

> This section will help you run ECK on Openshift

Reviewer: Missing final dot.

> This section will help you run ECK on Openshift.

Reviewer 2: > This section explains how to run ECK on Openshift.

> This section will help you to run the quickstart on Openshift.

Reviewer 3: I don't think the term 'quickstart' would be understood by everyone
```suggestion
This section will help you run ECK on Openshift
```

> NOTE: Only Elasticsearch and Kibana are supported on Openshift or with a secured cluster for the moment.

Reviewer 4: Can you clarify what this means? The "or with a secured cluster" doesn't have a clear meaning to me at least

Author: You are right. I have removed it because it is more misleading than helpful. K8S clusters can be secured with PSP. Like the `restricted` SCC on Openshift, restricted PSP may prevent the deployment of the APM server _(which is not compatible with some of these policies)_

=== Prerequisites

> . To run the following instructions, you must be a `system:admin` user or a User with the privileges to create Projects, CRDs and RBACs resources at the cluster level.

Reviewer 4: ```suggestion
. To run the following instructions, you must be a `system:admin` user or a User with the privileges to create Projects, CRDs, and RBAC resources at the cluster level.
```

> . The following instructions must be run as the `system:admin` user or as a `User` with some privileges to create  `Projects`, `CRDs` and `RBACs` resources at the cluster level.

Reviewer 2: > To run the following instructions, you must be a `system:admin` user or a `User` with the privileges to create  `Projects`, `CRDs` and `RBACs` resources at the cluster level.

> . Set  virtual memory settings on the nodes

Reviewer: double space: `s|Set  virtual|Set virtual`

+

> Before deploying an Elasticsearch cluster with ECK you must ensure that the `vm.max_map_count` setting is correctly applied on all the nodes of your cluster. This is because the Pods created by ECK are likely to run with the `restricted` SCC, they will run with a limited set of privileges and will not be able to tune this setting. You can refer to the following https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html[documentation] for more information.

Reviewer 2: > Before deploying an Elasticsearch cluster with ECK, make sure you correctly applied the `vm.max_map_count` setting on all the nodes of your cluster. You will not be able to tune this setting afterwards, as the Pods created by ECK often run with a limited set of privileges (`restricted` SCC). Check the https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html[Virtual memory documentation] for more information.

Reviewer 2: It may be worth to spell the SCC acronym - I searched for it but couldn't find anything relevant.

> Before deploying an Elasticsearch cluster with ECK you must ensure that the `vm.max_map_count` setting is correctly applied on all the nodes of your cluster. This is because the Pods created by ECK are likely to run with the `restricted` SCC, they will run with a limited set of privileges and will not be able to tune this setting. You can refer to the following https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html[documentation] for more information.

Reviewer 4: ```suggestion
Before deploying an Elasticsearch cluster with ECK, you must ensure that the `vm.max_map_count` setting is correctly applied on all the nodes of your cluster. This is because the Pods created by ECK are likely to run with the `restricted` SCC: they will run with a limited set of privileges and will not be able to change this setting on the nodes that host them. You can refer to the following https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html[documentation] for more information.
```

> Before deploying an Elasticsearch cluster with ECK you have to ensure that the `vm.max_map_count` setting is correctly set on all the nodes of your cluster. This is because the Pods created by ECK are likely to run with the `restricted` SCC, they will run with a limited set of privileges and will not be able to tune this setting. You can refer to the following https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html[documentation] for more information.

Reviewer 4: ```suggestion
Before deploying an Elasticsearch cluster with ECK you must ensure that the `vm.max_map_count` setting is correctly applied on all the nodes of your cluster. This is because the Pods created by ECK are likely to run with the `restricted` SCC, they will run with a limited set of privileges and will not be able to tune this setting. You can refer to the following https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html[documentation] for more information.
```

=== Deploy the operator

> . Apply the all in one template as described in the quickstart

Reviewer 2: Apply the all-in-one template as described in the quickstart.

> . Apply the all-in-one template as described in the quickstart.

Reviewer 4: Might be worth linking to the quickstart here just in case we forget to keep this up to date

+
[source,shell]
----
oc apply -f https://download.elastic.co/downloads/eck/0.9.0/all-in-one.yaml
----

> . [Optional] If the Software Defined Network is configured with the `ovs-multitenant` plug-in you need to allow the `elastic-system` namespace to access other pods and services in the cluster: 

Reviewer 2: -> Pods (capital P)
-> insert a comma before "you need"

> . [Optional] If the Software Defined Network is configured with the `ovs-multitenant` plug-in you need to allow the `elastic-system` namespace to access other pods and services in the cluster: 

Reviewer 4: ```suggestion
. [Optional] If the Software Defined Network is configured with the `ovs-multitenant` plug-in, you need to allow the `elastic-system` namespace to access other pods and services in the cluster: 
```

+
[source,shell]
----
oc adm pod-network make-projects-global elastic-system
----

> . Create a namespace that will hold the Elastic resources (Elasticsearch, Kibana):

Reviewer 2: -> that holds
avoid Future tense wherever possible

Reviewer 4: In this case it does not hold the resources yet though, so it seems fine? Though I might also suggest "to hold the Elastic..."

+
[source,shell]
----
oc new-project elastic # creates the elastic project
----
+

> By default the operator will watch and create resources in the `default` namespace. You need to patch the operator to manage resources in an other namespace.

Reviewer 4: Might be worth providing an example of listening cluster wide here as well, since I imagine that is a common use case

Author: I would be tempted to say that it is not specific to Openshift, so it should be documented elsewhere. That said it's an interesting point, I'm not sure we have a document explaining how to do that.

Reviewer 4: Yeah that's a good point. We should definitely add that to a larger configuration doc and perhaps link to it from here

> By default the operator will watch and create resources in the `default` namespace. You need to patch the operator to manage resources in an other namespace.

Reviewer 2: Remove Future tense.

> By default the operator will watch and create resources in the `default` namespace. You need to patch the operator to manage resources in an other namespace.

Reviewer 4: ```suggestion
By default the operator watches and creates resources in the `default` namespace. You can patch the operator to manage resources in another namespace:
```

+
[source,shell]
----
kubectl patch statefulset/elastic-operator \
  -n elastic-system \ 
  --type='json' \
  --patch '[{"op":"add","path":"/spec/template/spec/containers/0/env/-","value": {"name": "NAMESPACE", "value": "elastic"}}]'
----
+
Replace `elastic` in the examples above with the name of the namespace in which you want to deploy some resources.

. [Optional] If you want to allow an other user or a group of users to manage the Elastic resources:
+
[source,shell]
----
oc adm policy add-role-to-user elastic-operator developer -n elastic
----
+
In the example above the user `developer` is allowed to manage Elastic resources in the namespace `elastic`.

=== Deploy an Elasticsearch instance with a route

> The following command creates an Elasticsearch cluster `elasticsearch-sample` and a "passthrough" route to access it:

Reviewer 2: Use the following command to create an Elasticsearch cluster....

[source,shell]
----
cat <<EOF | oc apply -n elastic -f -
# This sample sets up an Elasticsearch cluster with an OpenShift route
apiVersion: elasticsearch.k8s.elastic.co/v1alpha1
kind: Elasticsearch
metadata:
  name: elasticsearch-sample
spec:
  version: "7.2.0"
  setVmMaxMapCount: false
  nodes:
  - config:
      node.master: true
      node.data: true
    nodeCount: 1
---
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  name: elasticsearch-sample
spec:
  #host: elasticsearch.example.com # override if you don't want to use the host that is automatically generated by OpenShift (<route-name>[-<namespace>].<suffix>)
  tls:
    termination: passthrough # Elasticsearch is the TLS endpoint
    insecureEdgeTerminationPolicy: Redirect
  to:
    kind: Service
    name: elasticsearch-sample-es-http
EOF
----

=== Deploy a Kibana instance with a route

> In the same way the following command creates a Kibana instance and a "passthrough" route to access it:

Reviewer 2: Use the following command to create a Kibana instance...

[source,shell]
----
cat <<EOF | oc apply -n elastic -f -
apiVersion: kibana.k8s.elastic.co/v1alpha1
kind: Kibana
metadata:
  name: kibana-sample
spec:
  version: "7.2.0"
  nodeCount: 1
  elasticsearchRef:
    name: "elasticsearch-sample"
  podTemplate:
    spec:
      containers:
      - name: kibana
        resources:
          limits:
            memory: 1Gi
            cpu: 1
---
apiVersion: v1
kind: Route
metadata:
  name: kibana-sample
spec:
  #host: kibana.example.com # override if you don't want to use the host that is automatically generated by OpenShift (<route-name>[-<namespace>].<suffix>)
  tls:
    termination: passthrough # Kibana is the TLS endpoint
    insecureEdgeTerminationPolicy: Redirect
  to:
    kind: Service
    name: kibana-sample-kb-http
EOF
----

> You can get the hosts of each `Routes` with the following command:

Reviewer 2: Use the following command to get the hosts of each `Routes`:

> You can get the hosts of each `Routes` with the following command:

Reviewer 4: ```suggestion
Use the following command to get the hosts of each `Route`:
```
